### PR TITLE
NVDIMM plugin

### PR DESCRIPTION
**Pull request by vojtechtrefny**
New plugin for working with NVDIMM devices. It contains only some basic functionality for namespace management.
This functionality is mostly needed for [blivet](https://github.com/storaged-project/blivet). I'm marking this as WIP for now, because I'll probably need to add more functions and/or change some of the functions below later, based on the blivet implementation.

### CODE REVIEW

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

File: src/plugins/nvdimm.c [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+    ndctl_unref (ctx);
+    return TRUE;
+}
+
+static BDNVDIMMNamespaceInfo* get_nvdimm_namespace_info (struct ndctl_namespace *ndns, GError **error) {
+    struct ndctl_btt *btt;
+    struct ndctl_pfn *pfn;
+    struct ndctl_dax *dax;
+    enum ndctl_namespace_mode mode;
+    gchar buf[37] = {0};
```
Comment (vpodzime) [inline src/plugins/nvdimm.c]: Why ``37``?

File: src/plugins/nvdimm.c [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+            info->size = ndctl_btt_get_size (btt);
+            info->mode = BD_NVDIMM_NAMESPACE_MODE_SECTOR;
+            break;
+        case NDCTL_NS_MODE_RAW:
+            info->size = ndctl_namespace_get_size (ndns);
+            info->mode = BD_NVDIMM_NAMESPACE_MODE_RAW;
+            break;
+        default:
+            g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                         "Failed to get information about namespaces.");
```
Comment (vpodzime) [inline src/plugins/nvdimm.c]: I think that adding the mode to the error message is going to simplify debugging one day.

File: src/plugins/nvdimm.c [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        return NULL;
+    }
+
+    /* now create the return value -- NULL-terminated array of BDNVDIMMNamespaceInfo */
+    infos = g_new0 (BDNVDIMMNamespaceInfo*, namespaces->len + 1);
+    for (i=0; i < namespaces->len; i++)
+        infos[i] = (BDNVDIMMNamespaceInfo*) g_ptr_array_index (namespaces, i);
+    infos[i] = NULL;
+
+    g_ptr_array_free (namespaces, FALSE);
```
Comment (vpodzime) [inline src/plugins/nvdimm.c]: With ``g_ptr_array_free (namespaces, FALSE)`` there's no need to create the return value with the ``for`` cycle above, the inner contents of the ``GPointerArray`` can be used as is.

File: src/plugins/nvdimm.c [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+
+    if (!check_deps (&avail_deps, DEPS_NDCTL_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+      return FALSE;
+
+    mode_str = bd_nvdimm_namespace_get_mode_str (mode, error);
+    if (!mode_str)
+        /* error is already populated */
+        return FALSE;
+
+    args[3] = g_strdup (namespace);
```
Comment (vpodzime) [inline src/plugins/nvdimm.c]: Why not use ``namespace`` directly?

File: tests/nvdimm_test.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        # skip the test if there is more than one device
+        # these tests change nvdimm mode a we need to be sure that we are really
+        # working with the 'fake' device created by the memmap kernel cmdline option
+        if type(self.sys_info) is not dict:
+            self.skipTest("Multiple NVDIMM devices found.")
+
+        # skip the tests if the nvdimm is a 'fake' one
+        cmdline = read_file("/proc/cmdline")
+        if "memmap=" not in cmdline:
+            self.skipTest("NVDIMM device found but not created by the 'memmap' kernel command-line option.")
```
Comment (vpodzime) [inline tests/nvdimm_test.py]: *"found, but..."*

**Pull request merged**